File: p304m/num_matrix.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <iostream>
#include <vector>

using namespace std;

class NumMatrix {
    vector<vector<int>> sums; // (rows + 1) * (cols + 1)
public:
    NumMatrix(vector<vector<int>> matrix)
        : sums(matrix.size() + 1,
            vector<int>(matrix[0].size() + 1, 0))
    {
        for (int r = 0; r < matrix.size(); ++r) {
            vector<int> rowsum(matrix[r].size() + 1, 0);
            for (int c = 0; c < matrix[r].size(); ++c) {
                rowsum[c + 1] = rowsum[c] + matrix[r][c];
                sums[r + 1][c + 1] = sums[r][c + 1] + rowsum[c + 1];
            }
        }
    }

    int sumRegion(int row1, int col1, int row2, int col2)
    {
        return sums[row2 + 1][col2 + 1]
            - sums[row2 + 1][col1]
            - sums[row1][col2 + 1]
            + sums[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */
int main(int argc, char* argv[])
{
    NumMatrix obj(
        { { 3, 0, 1, 4, 2 },
            { 5, 6, 3, 2, 1 },
            { 1, 2, 0, 1, 5 },
            { 4, 1, 0, 1, 7 },
            { 1, 0, 3, 0, 5 } });
    cout << "sumRegion(2, 1, 4, 3) -> " << obj.sumRegion(2, 1, 4, 3) << endl;
    cout << "sumRegion(1, 1, 2, 2) -> " << obj.sumRegion(1, 1, 2, 2) << endl;
    cout << "sumRegion(1, 2, 2, 4) -> " << obj.sumRegion(1, 2, 2, 4) << endl;
    return 0;
}

File: p304m/num_matrix.py
from typing import List
from itertools import accumulate


class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        m, n = len(matrix), len(matrix[0])
        self.psum = [[0] * (n + 1)] + [[0] + list(accumulate(row)) for row in matrix]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                self.psum[i][j] += self.psum[i - 1][j]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return (
            self.psum[row2 + 1][col2 + 1]
            - self.psum[row1][col2 + 1]
            - self.psum[row2 + 1][col1]
            + self.psum[row1][col1]
        )


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
nm = NumMatrix(
    [
        [3, 0, 1, 4, 2],
        [5, 6, 3, 2, 1],
        [1, 2, 0, 1, 5],
        [4, 1, 0, 1, 7],
        [1, 0, 3, 0, 5],
    ]
)
for region, expected in [([2, 1, 4, 3], 8), ([1, 1, 2, 2], 11), ([1, 2, 2, 4], 12)]:
    row1, col1, row2, col2 = region
    actual = nm.sumRegion(row1, col1, row2, col2)
    print(f"sumRegion({row1}, {col1}, {row2}, {col2}) -> {actual}")

